File: Source/hardware.java
/**
 * Static-Singleton-Hardwareklasse für CalliopeIO,
 * analog zur Pico-IO-Variante.
 *
 * Beispiel-Aufruf in BlueJ:
 *      hardware.ledOn(2, 2);
 *      hardware.scrollText("Hallo");
 *      hardware.setRgbAll(255, 0, 0);
 *
 * Die Verbindung zum Calliope mini wird automatisch beim ersten Zugriff aufgebaut.
 */
public final class hardware {

    private static CalliopeIO calliope = null;

    private static synchronized CalliopeIO get() {
        if (calliope == null) calliope = new CalliopeIO();
        return calliope;
    }

    // ---- LED-Matrix --------------------------------------------------------
    public static void ledOn(int x, int y)                     { get().ledOn(x, y); }
    public static void ledOff(int x, int y)                    { get().ledOff(x, y); }
    public static void ledSet(int x, int y, boolean state)     { get().ledSet(x, y, state); }
    public static void ledDim(int x, int y, int brightness)    { get().ledDim(x, y, brightness); }
    public static void clearDisplay()                          { get().clearDisplay(); }
    public static void showImage(int[][] grid)                 { get().showImage(grid); }
    public static void scrollText(String s)                    { get().scrollText(s); }
    public static void scrollText(String s, int delay)         { get().scrollText(s, delay); }
    public static void scrollNumber(int n)                     { get().scrollNumber(n); }
    public static void scrollNumber(int n, int delay)          { get().scrollNumber(n, delay); }
    public static void enableDisplay(boolean enable)           { get().enableDisplay(enable); }

    // ---- RGB-LEDs ----------------------------------------------------------
    public static void setRgb(int led, int r, int g, int b)    { get().setRgb(led, r, g, b); }
    public static void setRgbAll(int r, int g, int b)          { get().setRgbAll(r, g, b); }
    public static void clearRgb()                              { get().clearRgb(); }

    // ---- Motoren -----------------------------------------------------------
    public static void setMotor(int m, boolean fwd, int speed) { get().setMotor(m, fwd, speed); }
    public static void stopMotor(int m)                        { get().stopMotor(m); }
    public static void stopMotors()                            { get().stopMotors(); }

    // ---- Buttons -----------------------------------------------------------
    public static boolean isButtonAPressed()                   { return get().isButtonAPressed(); }
    public static boolean isButtonBPressed()                   { return get().isButtonBPressed(); }
    public static boolean wasButtonAPressed()                  { return get().wasButtonAPressed(); }
    public static boolean wasButtonBPressed()                  { return get().wasButtonBPressed(); }
    public static void clearPressed()                          { get().clearPressed(); }

    // ---- Sensoren ----------------------------------------------------------
    public static int getAccelerationX()                       { return get().getAccelerationX(); }
    public static int getAccelerationY()                       { return get().getAccelerationY(); }
    public static int getAccelerationZ()                       { return get().getAccelerationZ(); }
    public static int getCompassX()                            { return get().getCompassX(); }
    public static int getCompassY()                            { return get().getCompassY(); }
    public static int getCompassZ()                            { return get().getCompassZ(); }
    public static int getCompassHeading()                      { return get().getCompassHeading(); }
    public static int getTemperature()                         { return get().getTemperature(); }
    public static int getLightLevel()                          { return get().getLightLevel(); }
    public static boolean wasShaken()                          { return get().wasShaken(); }

    // ---- Ringpads ----------------------------------------------------------
    public static void setPadAnalogIn(int pad)                 { get().setPadAnalogIn(pad); }
    public static int  readPadAnalog(int pad)                  { return get().readPadAnalog(pad); }
    public static void setPadDigitalIn(int pad, boolean pull)  { get().setPadDigitalIn(pad, pull); }
    public static boolean readPadDigital(int pad)              { return get().readPadDigital(pad); }
    public static void setPadDigitalOut(int pad)               { get().setPadDigitalOut(pad); }
    public static void writePadDigital(int pad, boolean s)     { get().writePadDigital(pad, s); }
    public static void setPadPwm(int pad)                      { get().setPadPwm(pad); }
    public static void writePadPwm(int pad, int v)             { get().writePadPwm(pad, v); }

    // ---- Lautsprecher ------------------------------------------------------
    public static void playBeep(int duration)                  { get().playBeep(duration); }
    public static void playTone(int frequency, int duration)   { get().playTone(frequency, duration); }
    public static void stopBeep()                              { get().stopBeep(); }

    // ---- Loop --------------------------------------------------------------
    public static int loop(Object o, String m)                 { return get().startLoop(o, m); }
    public static int loopWithDelay(Object o, String m, int d) { return get().startLoopWithDelay(o, m, d); }
    public static void stopLoop(int id)                        { get().stopLoop(id); }
    public static void stopAllLoops()                          { get().stopAllLoops(); }
    public static void listLoops()                             { get().listLoops(); }

    // ---- Timing ------------------------------------------------------------
    public static void pause(int ms)                           { get().pause(ms); }
    public static void startClock()                            { get().startClock(); }
    public static long getClock()                              { return get().getClock(); }
    public static long stopClock()                             { return get().stopClock(); }

    // ---- Connection --------------------------------------------------------
    public static boolean getStatus()                          { return get().getStatus(); }
    public static void reconnect()                             { calliope = new CalliopeIO(); }
    public static void close()                                 { get().close(); }
}

File: Software/Python/Source/hardware.py
"""Modul-Wrapper für CalliopeIO — Python-Pendant zur Java-Klasse ``hardware``.

Beispiel::

    import hardware
    hardware.led_on(2, 2)
    hardware.scroll_text("Hallo")
    hardware.set_rgb_all(0, 255, 0)
    print(hardware.temperature(), "°C")

Die Verbindung wird beim ersten Aufruf automatisch hergestellt.
"""

from __future__ import annotations

import threading
from typing import Callable, Sequence

from calliope_io import CalliopeIO


_calliope: CalliopeIO | None = None
_lock = threading.Lock()


def _io() -> CalliopeIO:
    global _calliope
    with _lock:
        if _calliope is None:
            _calliope = CalliopeIO()
        return _calliope


# ---- LED-Matrix ------------------------------------------------------------
def led_on(x: int, y: int) -> None:                                _io().led_on(x, y)
def led_off(x: int, y: int) -> None:                               _io().led_off(x, y)
def led_set(x: int, y: int, state: bool) -> None:                  _io().led_set(x, y, state)
def led_dim(x: int, y: int, brightness: int) -> None:              _io().led_dim(x, y, brightness)
def clear_display() -> None:                                       _io().clear_display()
def show_image(grid: Sequence[Sequence[int]]) -> None:             _io().show_image(grid)
def scroll_text(text: str, delay: int = 120) -> None:              _io().scroll_text(text, delay)
def scroll_number(n: int, delay: int = 120) -> None:               _io().scroll_number(n, delay)
def enable_display(enable: bool) -> None:                          _io().enable_display(enable)

# ---- RGB-LEDs --------------------------------------------------------------
def set_rgb(led: int, r: int, g: int, b: int) -> None:             _io().set_rgb(led, r, g, b)
def set_rgb_all(r: int, g: int, b: int) -> None:                   _io().set_rgb_all(r, g, b)
def clear_rgb() -> None:                                           _io().clear_rgb()

# ---- Motoren ---------------------------------------------------------------
def set_motor(motor: int, forward: bool, speed: int) -> None:      _io().set_motor(motor, forward, speed)
def stop_motor(motor: int) -> None:                                _io().stop_motor(motor)
def stop_motors() -> None:                                         _io().stop_motors()

# ---- Buttons ---------------------------------------------------------------
def is_button_a_pressed() -> bool:                          return _io().is_button_a_pressed()
def is_button_b_pressed() -> bool:                          return _io().is_button_b_pressed()
def was_button_a_pressed() -> bool:                         return _io().was_button_a_pressed()
def was_button_b_pressed() -> bool:                         return _io().was_button_b_pressed()
def clear_pressed() -> None:                                       _io().clear_pressed()

# ---- Sensoren --------------------------------------------------------------
def acceleration_x() -> int:                                return _io().acceleration_x()
def acceleration_y() -> int:                                return _io().acceleration_y()
def acceleration_z() -> int:                                return _io().acceleration_z()
def acceleration() -> tuple[int, int, int]:                 return _io().acceleration()
def compass_x() -> int:                                     return _io().compass_x()
def compass_y() -> int:                                     return _io().compass_y()
def compass_z() -> int:                                     return _io().compass_z()
def compass() -> tuple[int, int, int]:                      return _io().compass()
def compass_heading() -> int:                               return _io().compass_heading()
def temperature() -> int:                                   return _io().temperature()
def light_level() -> int:                                   return _io().light_level()
def was_shaken() -> bool:                                   return _io().was_shaken()

# ---- Ringpads --------------------------------------------------------------
def set_pad_analog_in(pad: int) -> None:                           _io().set_pad_analog_in(pad)
def read_pad_analog(pad: int) -> int:                       return _io().read_pad_analog(pad)
def set_pad_digital_in(pad: int, pull_up: bool = False) -> None:   _io().set_pad_digital_in(pad, pull_up)
def read_pad_digital(pad: int) -> bool:                     return _io().read_pad_digital(pad)
def set_pad_digital_out(pad: int) -> None:                         _io().set_pad_digital_out(pad)
def write_pad_digital(pad: int, state: bool) -> None:              _io().write_pad_digital(pad, state)
def set_pad_pwm(pad: int) -> None:                                 _io().set_pad_pwm(pad)
def write_pad_pwm(pad: int, value: int) -> None:                   _io().write_pad_pwm(pad, value)

# ---- Lautsprecher ----------------------------------------------------------
def play_beep(duration_ms: int) -> None:                           _io().play_beep(duration_ms)
def play_tone(frequency: int, duration_ms: int) -> None:           _io().play_tone(frequency, duration_ms)
def stop_beep() -> None:                                           _io().stop_beep()

# ---- Loop ------------------------------------------------------------------
def loop(fn: Callable[[], None], delay_ms: int = 0) -> int: return _io().start_loop(fn, delay_ms)
def stop_loop(loop_id: int) -> None:                               _io().stop_loop(loop_id)
def stop_all_loops() -> None:                                      _io().stop_all_loops()
def list_loops() -> None:                                          _io().list_loops()

# ---- Timing ----------------------------------------------------------------
def pause(ms: int) -> None:                                        CalliopeIO.pause(ms)
def start_clock() -> None:                                         _io().start_clock()
def clock() -> int:                                         return _io().clock()
def stop_clock() -> int:                                    return _io().stop_clock()

# ---- Verbindung ------------------------------------------------------------
def is_connected() -> bool:                                 return _io().is_connected()
def open() -> None:                                                _io().open()
def close() -> None:                                               _io().close()
def reconnect() -> None:
    global _calliope
    with _lock:
        _calliope = CalliopeIO()
